### PR TITLE
handle exception, otherwise p.close() will hang

### PR DESCRIPTION
**Pull request by njzjz**
If an exception is raised in f, p.close() and p.join() will hang. We need p.terminate() to handle exception.

### CODE REVIEW

**Comment by blueyed** [comment]
Thanks.
Can this be added/covered in the tests?  (not really checked/compared them now)
(`test_multiprocessing_pool_terminate`, `test_multiprocessing_process_with_terminate`)

**Comment by njzjz** [comment]
Is it enough?

File: tests/test_pytest_cov.py [hunk 1]
```diff
@@ -1220,11 +1224,17 @@ def target_fn():
...
-    p.join()
+    try:
+        p.start()
+        time.sleep(0.5)
+        event.wait(1)
+    except:
+        p.terminate()
+        raise
+    else:
+        p.close()
```
Comment (ionelmc) [inline tests/test_pytest_cov.py]: There is no Process.close method?

**Comment by ionelmc** [comment]
> Is it enough?

I know we have some broken tests in certain pythons/configurations but new tests need to pass :)

File: tests/test_pytest_cov.py (hunk 1 above)
Comment (njzjz) [reply tests/test_pytest_cov.py]: https://docs.python.org/3/library/multiprocessing.html#multiprocessing.Process.close
It's just new in Python 3.7.

Comment (njzjz) [reply tests/test_pytest_cov.py]: I could comment this line, and enable it when all tests use python>=3.7 (maybe few years later...).

File: tests/test_pytest_cov.py [hunk 2]
```diff
@@ -1231,8 +1231,9 @@ def test_run_target():
...
         p.terminate()
         raise
-    else:
-        p.close()
+    # new in python 3.7
+    #else:
+    #    p.close()
     finally:
         p.join()
 ''')
```
Comment (blueyed) [inline tests/test_pytest_cov.py]: Enable/use it with `sys.version_info >= (3, 7)` then?

Comment (njzjz) [reply tests/test_pytest_cov.py]: Good point!

**Comment by njzjz** [comment]
> > Is it enough?
> 
> I know we have some broken tests in certain pythons/configurations but new tests need to pass :)

Now CI has passed on two functions that I modified. 😃

File: tests/test_pytest_cov.py [hunk 3]
```diff
@@ -1096,7 +1100,7 @@ def test_run_target():
     assert not testdir.tmpdir.listdir(".coverage.*")
     result.stdout.fnmatch_lines([
         '*- coverage: platform *, python * -*',
-        'test_multiprocessing_pool* 100%*',
+        'test_multiprocessing_pool* 99%*',
```
Comment (ionelmc) [inline tests/test_pytest_cov.py]: These assertions shouldn't be changed.

File: tests/test_pytest_cov.py [hunk 4]
```diff
@@ -1234,7 +1246,7 @@ def test_run_target():
 
     result.stdout.fnmatch_lines([
         '*- coverage: platform *, python * -*',
-        'test_multiprocessing_process* 16 * 100%*',
+        'test_multiprocessing_process* 22 * 8*%*',
```
Comment (ionelmc) [inline tests/test_pytest_cov.py]: Same here.

File: tests/test_pytest_cov.py (hunk 3 above)
Comment (njzjz) [reply tests/test_pytest_cov.py]: Yes, but I have no idea about how to skip L1082-1084 when calculating the coverage. https://github.com/pytest-dev/pytest-cov/blob/5f0e50398b801d6711480d63669a330d1ba3822a/tests/test_pytest_cov.py#L1082-L1084

**Pull request closed without merging**